[WIP] Add Tapas model

NielsRogge opened the pull request:
# What does this PR do?

Since the beginning of August, I'm working in my free time on incorporating the [Tapas](https://arxiv.org/abs/2004.02349) algorithm by Google AI in the Transformers library (because this library is awesome and I want to contribute to it!). Tapas is basically a BERT model with some clever modifications for natural language understanding related to **tabular data** (structured data like tables, or even HTML). Adding this model could foster research in this area 😄 

Demo's of my current implementation:
* [colab notebook](https://colab.research.google.com/drive/1feRe1Jyjtw7iZVRiKWulP6WjBW6hBIJE?usp=sharing) to showcase `TapasForQuestionAnswering` on WTQ (WikiTable Questions by Stanford University)
* [colab notebook](https://colab.research.google.com/drive/1CDPUr7c8uCNnCtAmmFfj91j-sIzdcqym?usp=sharing) to showcase `TapasForQuestionAnswering`on SQA (Sequential Question Answering by Microsoft Research)
* [colab notebook](https://colab.research.google.com/drive/1JDwWrwHSt8KhGBQ57BDlCFZEe0xMGTQA?usp=sharing) to showcase `TapasForSequenceClassification` on TabFact (Table Fact checking, introduced at ICLR this year)

The model weights are available on the [original Github repository](https://github.com/google-research/tapas), and I wrote a conversion script (similar to other models in the Transformers library) to load them into their PyTorch counterpart. 
I suggest reading the [paper](https://arxiv.org/abs/2004.02349) as well as my [notes](https://docs.google.com/document/d/1WIdZX6of1l-c4AmT909PT7Dpj57EfqUh8BBPaf9ztOw/edit?usp=sharing) to gain a full understanding of how the model works and how I implemented it. There's also a [blog post](https://ai.googleblog.com/2020/04/using-neural-networks-to-find-answers.html) by Google AI as well as a [video](https://www.youtube.com/watch?v=cIUtRNhY6Rw&ab_channel=YannicKilcher) by Yannic Kilcher explaining how the algorithm works. 

The main classes are `TapasConfig`, `TapasModel`, `TapasForQuestionAnswering` and `TapasForSequenceClassification` which can all be found in `modeling_tapas.py`. I'm quite sure the models are OK, the output is the same as the Tensorflow implementation. I added a very extensive documentation (docstrings) to all classes, which you can view by running the `make html` command from the docs directory. Feedback appreciated!

However, there are 2 things for which I need some help/opinions to finish this work:

## 1. Making TapasTokenizer fully Transformers-compliant
To implement `TapasTokenizer`, I need some help/opinions. I suggest using Pandas dataframes as the central object for tabular data (as shown in the Colab notebooks above). and let the API be as follows:

```
from transformers import TapasTokenizer
import pandas as pd

data = {'Actors': ["Brad Pitt", "Leonardo Di Caprio", "George Clooney"], 
        'Age': ["56", "45", "59"],
        'Number of movies': ["87", "53", "69"],
        'Date of birth': ["18 december 1963", "11 november 1974", "6 may 1961"]}
table = pd.DataFrame.from_dict(data)
queries = ["When was Brad Pitt born?", 
           "Which actor appeared in the least number of movies?",
           "What is the average number of movies?"]

tokenizer = TapasTokenizer.from_pretrained("tapas-base-finetuned-wtq")
inputs = tokenizer(table=table, queries=queries)
```

Currently I've only implemented the `batch_encode_plus` method of TapasTokenizer, because it's not really clear to me how to make it fully compatible with the Transformers library, since the way that data is prepared for the model is a bit different compared to BERT/RoBERTa/etc (see also my notes above). It's also not straightforward to make it compatible with the different padding/truncation strategies of Transformers. Currently, the way it works is as follows: there’s a function `_get_token_budget` in `tokenization_tapas.py` that calculates the number of tokens left for the flattened table after tokenizing a question. This is currently set to `self.model_max_length - (len(question_tokens) + 2)` (+ 2 for the CLS and SEP tokens), as was done in the original implementation. There is a hyperparameter when initializing `TapasTokenizer` called `drop_rows_to_fit` which drops rows of the table to fit into the token budget if set to `True`. If it’s set to `False` and a table is too big, it throws a `ValueError` indicating 'too many rows'.

## 2. Testing
Currently I've written `test_modeling_tapas.py` (23 tests passed, 5 failed) and `test_modeling_tapas_utilities.py` (9 tests passed). However, there are 4 different settings to use `TapasForQuestionAnswering` (see my notes) and these all need to be tested (currently only 1 setting is tested) - some help here would be great. Besides this, tests should be added to see whether the model can be properly trained, as well as adding `test_tokenization_tapas.py` (which depends on how TapasTokenizer will be implemented). 

Fixes the following issues (people requesting to add the model):
- #4166 
- #4288

## Who can review?
I suggest @sgugger @LysandreJik since we already discussed this on the forum [here](https://discuss.huggingface.co/t/adding-a-new-model-to-transformers-with-additional-dependencies/916/15).
 tokenizers: @mfuntowicz

DISCLAIMER: this is my first PR of my life, never done this before, hopefully I don't mess up anything (just got the Pro Git book 😛). I assume I should not use `git rebase` anymore now since this branch submitted as PR and should only use `git add`, `git commit` and `git push -u origin tapas_v3`? And `git pull origin tapas_v3` in case others make commits to my branch?
Is there a Slack channel where people can help me out in case I have git issues?

sgugger commented:
I think you might need a rebase on the latest master as your PR seems to have taken master from a while ago (all the modifications in the README should not be there for instance). If it messes the diff, we can always close this PR and open a new one, the branch will be safe :-)

NielsRogge commented:
I've always used `git fetch upstream` and `git rebase upstream/master`, before pushing my local `tapas_v3` branch to my fork on Github. When I now look at https://github.com/NielsRogge/transformers/tree/tapas_v3, it says "This branch is 15 commits ahead of huggingface:master.".

NielsRogge commented:
@sgugger should I keep rebasing this branch everyday, to keep up with master (as long as the code is not being reviewed)?

Also, is it normal that I have to do a force push everytime I perform a rebase and push to Github? Because when I want to do simply `git push -u origin tapas_v3`, I always get 
```
(env) PS C:\Users\niels.rogge\Documents\Python projecten\transformers> git push -u origin tapas_v3
To https://github.com/NielsRogge/transformers.git
 ! [rejected]          tapas_v3 -> tapas_v3 (non-fast-forward)
error: failed to push some refs to 'https://github.com/NielsRogge/transformers.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

after a rebase.

sgugger commented:
I'm far from being an expert on git and I don't use the command line anyway, so can't really help you with that.

LysandreJik commented on src/transformers/tokenization_tapas.py:
```diff
@@ -0,0 +1,1194 @@
+    return token.startswith("##")
+
+
+class TapasTokenizer(BertTokenizer):
```
I don't think this tokenizer needs to inherit from the `BertTokenizer`, as it behaves very differently.

LysandreJik reviewed (commented):
Hi! Thanks a lot @NielsRogge for implementing this model! It's a great model that definitely deserves its place in the huggingface library, and you did a great job implementing it! I'm reviewing the code below.

Regarding the tokenizer, we do not have a dependency (hard or soft) on pandas, but we will soon have one on `datasets`. I think it is feasible to have `datasets` as the central object for tabular data instead of pandas dataframes. It's very easy to use it as a replacement, your example code would simply use a `datasets.Dataset` instead of a `pd.Dataframe`:
```py
from transformers import TapasTokenizer
from datasets import Dataset

data = {'Actors': ["Brad Pitt", "Leonardo Di Caprio", "George Clooney"], 
        'Age': ["56", "45", "59"],
        'Number of movies': ["87", "53", "69"],
        'Date of birth': ["18 december 1963", "11 november 1974", "6 may 1961"]}

table = Dataset.from_dict(data)
queries = ["When was Brad Pitt born?", 
           "Which actor appeared in the least number of movies?",
           "What is the average number of movies?"]

tokenizer = TapasTokenizer.from_pretrained("tapas-base-finetuned-wtq")
inputs = tokenizer(table=table, queries=queries)
```
Putting @lhoestq in cc.

Regarding the configuration and tokenizer, it would be great to uncouple them from their BERT counterparts. 

I haven't played around with the model yet, but it seems to be in a great shape already. If you can do the modifications in the review here that would be great, and then I'll spend time playing with it and will come back to you regarding the tokenizer API currently setup.

Concerning git, it seems you have *just* rebased & force-pushed, so that's very cool. Regarding your question, this does mess up the history of users that are currently using your branch, but seeing as we're early in development, there shouldn't be any.

LysandreJik commented on src/transformers/modeling_tapas_utilities.py:
```diff
@@ -0,0 +1,855 @@
+# coding=utf-8
```
Same comment for the tokenizer utilities.

NielsRogge replied:
Ok, where is the best place to add everything of `modeling_tapas_utilities.py`: at the top of `modeling_tapas.py`, or before/after the forward() of `TapasForQuestionAnswering`? (more general question: does one always need to define functions before using them, or does this not matter?)

And I don't understand what you mean by the wrapper, you mean I have to use that wrapper around every torch.function in my `modeling_tapas.py` file? Do you want to make PyTorch a soft dependency?

LysandreJik replied:
I don't think we have a standard for where to put this, but as a general code readability observation I guess it would make sense to put the most important methods/classes higher up in the file, making them accessible quicker. Then it would mean putting the utilities at the bottom of the file, after the classes. What do you think, and what's your opinion on this @sshleifer, @patrickvonplaten, @sgugger?

No, I'm sorry I wasn't clear enough, I meant to say the `scatter` module, not PyTorch itself.

patrickvonplaten replied:
I agree with @LysandreJik, it would be awesome to have all this functionality in `modeling_tapas.py`. I would put these functions at the bottom of the file.

NielsRogge commented on src/transformers/configuration_tapas.py:
```diff
@@ -0,0 +1,154 @@
+TAPAS_PRETRAINED_CONFIG_ARCHIVE_MAP = {"tapas-base": "", "tapas-large": ""}  # to be added  # to be added
+
+
+class TapasConfig(BertConfig):
```
Done!

NielsRogge commented on src/transformers/modeling_tapas.py:
```diff
@@ -0,0 +1,1408 @@
+_CONFIG_FOR_DOC = "TapasConfig"
+_TOKENIZER_FOR_DOC = "TapasTokenizer"
+
+TAPAS_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "tapas-base",
+    "tapas-large",
+    # See all TAPAS models at https://huggingface.co/models?filter=tapas
+]
```
Yes, some notes about this:

* I would like to discuss which of the models to upload to the model hub, because the authors did release a large amount of model variants: https://github.com/google-research/tapas. I assume not all of them are that useful, especially the TINY models, and maybe we can only include the ones with relative embeddings (because performance is almost always slightly better with relative position embeddings) and intermediate pre-training (for the same reason). 

* The naming of the models is quite extensive: **tapas_wtq_wikisql_sqa_inter_masklm_large_reset** for example is the large model with relative position embeddings (reset) that was pretrained on masked language modeling, then intermediate pretraining, then fine-tuned on SQA, then on WikiSQL and finally on WTQ 😅 Is there a naming standard for models in the model hub? Maybe this can be shortened to **tapas-large-finetuned-wtq**, and provide the details in the model cards.

LysandreJik replied:
- Regarding the models to be added, I don't think it hurts to put all the models in the hub. This way it would be easy for future researchers to replicate the results of this paper, only using the implementation here, without needing to switch from one implementation to another. In any case, the conversion script should handle all of these.  We can take care of uploading the models then, as they'll need to be under the google organization namespace.

- I think your proposal regarding model naming makes total sense! As long as there exists a document in which we can find the old name, renaming to something easier to understand like this is fine.

patrickvonplaten commented on src/transformers/configuration_tapas.py:
```diff
@@ -0,0 +1,208 @@
+            The epsilon used by the layer normalization layers.
+        gradient_checkpointing (:obj:`bool`, `optional`, defaults to :obj:`False`):
+            If True, use gradient checkpointing to save memory at the expense of slower backward pass.
+        positive_weight (:obj:`float`, `optional`, defaults to 10.0):
```
(nit) think `positive_label_weight` is better here

NielsRogge replied:
Done

patrickvonplaten commented on src/transformers/configuration_tapas.py:
```diff
@@ -0,0 +1,208 @@
+            Weight for positive labels.
+        num_aggregation_labels (:obj:`int`, `optional`, defaults to 0):
+            The number of aggregation operators to predict.
+        aggregation_loss_importance (:obj:`float`, `optional`, defaults to 1.0):
```
also maybe replace `importance` with `weight`? => the name `weight` is more common IMO for multi-loss

NielsRogge replied:
Done

patrickvonplaten commented on src/transformers/configuration_tapas.py:
```diff
@@ -0,0 +1,208 @@
+        average_approximation_function: (:obj:`string`, `optional`, defaults to :obj:`"ratio"`):
+            Method to calculate expected average of cells in the relaxed case.
+        cell_select_pref: (:obj:`float`, `optional`, defaults to None):
+            Preference for cell selection in ambiguous cases.
```
This is a bit difficult to understand - could you maybe add one more line of explanation here? Why is a preference a `float` value?

NielsRogge replied:
Added additional documentation for better understanding

patrickvonplaten reviewed (commented):
This looks like an amazing contribution! I let @LysandreJik handle the next steps and am happy to do a final review at a later stage.

NielsRogge commented:
> Hi! Thanks a lot @NielsRogge for implementing this model! It's a great model that definitely deserves its place in the huggingface library, and you did a great job implementing it! I'm reviewing the code below.

Thank you! Great to hear :)

I addressed all of the comments. To summarize:
* `TapasConfig` and `TapasTokenizer` now inherit from `PretrainedConfig` and `PreTrainedTokenizer` respectively. For `TapasTokenizer`, a lot of the code of `tokenization_bert.py` was copied (such as the `BasicTokenizer` and `WordPieceTokenizer` classes), since the tokenization logic of text itself is the same. However, some things are different (see `tokenization_tapas.py`). 
* `Modeling_tapas_utilities.py` and `tokenization_tapas_utilities.py` are also gone now, they are added to the bottom of `modeling_tapas.py` and `tokenization_tapas.py` respectively. 
* `pandas` is not a dependency in the code (`pandas` is not imported in any of the files), but I assume you want to switch to `datasets` so that people can use Tapas using only the `Transformers` library? However, currently, in `tokenization_tapas.py` some Pandas logic is used, for example in the `_tokenize_table` function `.iterrows()` is used, so this will require some changes.
* concerning git, I assume I should stop rebasing at some point? I can do it as long as I'm the only one committing?

LysandreJik commented on tests/test_modeling_tapas_utilities.py:
```diff
@@ -0,0 +1,202 @@
+        np.testing.assert_array_equal(result.numpy(), [[[1, 2], [3, 4]], [[7, 8], [5, 6]]])
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Don't think we need this

LysandreJik commented on tests/test_modeling_tapas.py:
```diff
@@ -0,0 +1,391 @@
+        TapasForQuestionAnswering,
+    )
+
+global_rng = random.Random()
+
+
+def token_type_ids_tensor_tapas(shape, type_vocab_sizes, rng=None, name=None):
+    #  Creates a random int32 tensor of the shape within the type vocab sizes
+    if rng is None:
+        rng = global_rng
+
+    values = []
+    for example in range(shape[0]):
+        for token in range(shape[1]):
+            for idx in range(shape[2]):
+                values.append(rng.randint(0, type_vocab_sizes[idx] - 1))
+
+    return torch.tensor(data=values, dtype=torch.long, device=torch_device).view(shape).contiguous()
```
Shouldn't need to use this. You can use the `ids_tensor` you have imported previously to generate random values of a given shape, bound to certain values.

NielsRogge replied:
Fixed, now using `ids_tensor` function.

LysandreJik commented on tests/test_modeling_tapas.py:
```diff
@@ -0,0 +1,391 @@
+# coding=utf-8
+# Copyright 2018 XXX Authors.
```
Should put the authors here.

LysandreJik commented on src/transformers/configuration_tapas.py:
```diff
@@ -0,0 +1,210 @@
+# coding=utf-8
+# Copyright (...) and The HuggingFace Inc. team.
```
Copyright should say Google Research

LysandreJik commented on src/transformers/configuration_tapas.py:
```diff
@@ -0,0 +1,210 @@
+
+    Args:
+        vocab_size (:obj:`int`, `optional`, defaults to 30522):
+            Vocabulary size of the BERT model. Defines the number of different tokens that can be represented by the
+            :obj:`inputs_ids` passed when calling :class:`~transformers.BertModel` or
+            :class:`~transformers.TFBertModel`.
```
```suggestion
            Vocabulary size of the TaPas model. Defines the number of different tokens that can be represented by the
            :obj:`inputs_ids` passed when calling :class:`~transformers.TapasModel`.
```

LysandreJik commented on src/transformers/configuration_tapas.py:
```diff
@@ -0,0 +1,210 @@
+            Delta parameter used to calculate the regression loss.
+        temperature: (:obj:`float`, `optional`, defaults to 1.0):
+            Scales cell logits to control the skewness of probabilities.
+        agg_temperature: (:obj:`float`, `optional`, defaults to 1.0):
```
Would advocate for replacing all shortened terms for their complete definition: `agg` > `aggregation`.
Same with `pref` > `preference`

LysandreJik commented on src/transformers/modeling_tapas.py:
```diff
@@ -0,0 +1,2222 @@
+# coding=utf-8
+# Copyright (...) and The HuggingFace Inc. team.
```
Google Research

LysandreJik reviewed (commented):
I've played around with your implementation, and it works well! It's very cool. Now comes a complicated aspect: ensuring we stay consistent with the existing API, especially for the tokenizers.

This is a complicated part, so please let us know if you would like some help along the way/if you want us to take over from here, in which case we would open PRs against your branch with proposals.

### Tokenizers

Right now, the tokenizer is not behaving as we would like as it doesn't respect much of the flags passed to it. For example, running your example:

```py
from transformers import TapasTokenizer

dir_name = "./tapas_base_finetuned_wtq_with_reset_and_intermediate_pretraining"

tokenizer = TapasTokenizer(vocab_file=dir_name + "/vocab.txt", model_max_length=512)
encoded_inputs = tokenizer.batch_encode_plus(table=table, queries=queries)
```
This immediately pads to the maximum length. This is the case whether I specify the `padding` strategy or not. It does not respect the `truncation` parameter either, nor does it respect the `return_attention_mask`, but it does respect the `return_token_type_ids`.

I haven't tried the `pad_to_multiple_of`, `is_split_into_words`, `return_overflowing_tokens`, `return_special_tokens_mask`, `return_length` parameters, but having them work similarly to other tokenizers should be a priority.

#### `__call__`

The `__call__` method should be implemented for the tokenizer. The following methods should also be implemented:

:warning: some of these are already implemented in the superclass, and may not need to be reimplemented. This may be the case for `prepare_for_model`, which takes care of the truncation/padding on its own.

- `build_inputs_with_special_tokens`
- `get_special_tokens_mask`
- `create_token_type_ids_from_sequences`

It seems you've only implemented the batching methods, but there are methods to be used for single sequences as well:

- `encode_plus`
- `prepare_for_model`

### Testing 

#### Modeling Tapas Utilities

Same as the other utilities, it would be nice to have this in the main file. Moving `test_modeling_tapas_utilities.py` to be in the same file as `test_modeling_tapas.py` would be great.

#### Tokenization tests

The implementation of the tokenization tests would be a good way to monitor what is currently working/failing.

NielsRogge commented on src/transformers/tokenization_tapas.py:
```diff
@@ -0,0 +1,2053 @@
+        # Two extra spots of SEP and CLS.
+        return len(question_tokens) + 2
+
+    def _get_token_budget(self, question_tokens):
```
Added

NielsRogge commented:
Hey @LysandreJik, thank you for your feedback.

I've fixed all comments that you had, apart from the tokenizer itself. 

> This is a complicated part, so please let us know if you would like some help along the way/if you want us to take over from here, in which case we would open PRs against your branch with proposals.

I'm happy to accept PRs against my branch, because it's not that clear to me how the tokenizer should be implemented in the best possible way.

LysandreJik commented:
Alright, I'll take a look and open a PR on your fork with the proposed changes. Yes, please don't rebase on `master` anymore as it would mess up my history as I start working on your branch.

We can keep it as it is until we merge now, and fix the merge conflict as the last step.

LysandreJik commented:
Hi @NielsRogge, I'm nearly done with the tokenizer changes, but we're focusing on getting version v3.5.0 out today and tomorrow. I'll try to open a PR on your repository then. Please hold off from adding commits now, as rebasing/merging would be very painful now! :smile:

NielsRogge commented:
@LysandreJik no worries, I'm not working on the `tapas_v3` branch.

However, there's still some important work left to do in terms of preparing the data for the model. In the original implementation, this is done in 3 steps:

## 1. TSV in the SQA format
Any dataset (SQA, WTQ, WikiSQL) is first transformed into a TSV with the same columns as the SQA format:
* position: integer indicating if the question is the first, second, third,... related to the table. Only required in case of conversational setup (such as SQA)
* question: string
* table_file: string, name of a csv file containing the tabular data
* answer_text: list of strings (each string being a cell value that is part of the answer)
* answer_coordinates: list of string tuples (each tuple being a cell coordinate, i.e. row, column pair that is part of the answer)
* aggregation_label: only required in case of strong supervision for aggregation (such as WikiSQL-supervised)
* answer_float: float answer to the question. Only required in case of weak supervision for aggregation (such as WTQ and WikiSQL)

If people want to fine-tune `TapasForQuestionAnswering` on their own dataset, they must prepare it in this TSV format, and associated csv files containing the tabular data. It would be great if we can upload all datasets (SQA, WTQ, WikiSQL and WikiSQL-supervised) in SQA format to the HuggingFace datasets hub.

## 2. Intermediate format: Interaction
Next, each table-question pair is transformed into an intermediate protocol buffer message which the authors call **Interaction**. Its properties are defined [here](https://github.com/google-research/tapas/blob/master/tapas/protos/interaction.proto), and include things like Table, Question, Answer, AnswerCoordinate, Cell, NumericValue, NumericValueSpan, etc. Populating all the fields of an Interaction based on the TSV is defined in [interaction_utils.py](https://github.com/google-research/tapas/blob/master/tapas/utils/interaction_utils.py), [interaction_utils_parser.py](https://github.com/google-research/tapas/blob/master/tapas/utils/interaction_utils_parser.py), [number_annotation_utils.py](https://github.com/google-research/tapas/blob/master/tapas/utils/number_annotation_utils.py), [number_utils.py](https://github.com/google-research/tapas/blob/master/tapas/utils/number_utils.py) and [text_utils.py](https://github.com/google-research/tapas/blob/master/tapas/utils/text_utils.py).

## 3. tf.train.Example
Finally, each interaction is transformed into an actual training example (`tf.train.Example`), containing the input_ids, mask, etc. as `tf.train.Feature` objects. This is is defined in [tf_example_utils.py](https://github.com/google-research/tapas/blob/master/tapas/utils/tf_example_utils.py).

_________________
The remaining work is basically step 2. As I worked with `Pandas` as standard format for tables in my implementation, my idea was to remove the need for the intermediate Interaction format, and instead define regular Python classes for each property of the Interaction proto to directly turn a row in the TSV file into a dictionary with PyTorch tensors as values. That is why I have defined a `NumericValue` class, `NumericValueSpan` class, `Cell` class, `Date` class, etc. in `tokenization_tapas.py`. 

I've noticed that the creation of the numeric values is not entirely correct in the `tapas_v3` branch. I'm now working on a correct implementation of this in a branch called `tapas_v3_up_to_date_with_master` (in which I also regularly rebase with upstream/master). It only involves changes to `tokenization_tapas.py`. The changes can eventually be added to `tapas_v3`. 

So my questions are:
- for each of the `xxx_utils.py` files which are used in step 2, there are corresponding `xxx_utils_test.py` files. Could you help in setting up tests in `tokenization_tapas_test.py`, to make sure we're following the original implementation? 
- I'm still assuming that tables are `pandas` dataframes in `tokenization_tapas.py`. Is this OK? Or do you want to change to `dataset`? Wouldn't it be more logical to have SQA/WTQ/WikiSQL as `dataset` objects, and the actual tables as `pandas` dataframes? Pandas is not a dependency of `tokenization_tapas.py`, but tables must be provided as a Pandas dataframe to `TapasTokenizer`.

LysandreJik commented:
Hi @NielsRogge! I just finished the tokenizer and its tests. The tests were kind of painful, as the API is a bit different (accepting a dataframe instead of a string), so I had to override most of the tests.

Here's how you can review:

- I did a `make style` on your branch, as I find the code easier to navigate once it's on par with our style.
- However, this introduces a bunch of changes that would make the PR hard to review.
- In order to circumvent this, I've pushed two branches: 
    - `tapas-style`, which is the exact branch you have, but with `make style` run on it and a few cosmetic adjustments
    - `tapas-final`, which builds on top of `tapas-style` to implement all the tokenizer API and tests
- For you to review, the easiest would be to review the PR I opened [here](https://github.com/huggingface/transformers/pull/8482), which aims to merge `tapas-final` into `tapas-style`. This way you can see the actual changes and only these.
- I described my changes in that PR's description so as not to clog this one.
- I setup a todo list if items remaining on the tokenizer, which are not blocking for the merge.

Please review https://github.com/huggingface/transformers/pull/8482, and tell me if you're okay with the changes. If you're okay, I'll merge `tapas-final` into `tapas-style`, and open a PR on your fork with the branch `tapas-style`, which will have all the changes.

This is one of the best hands-on introduction to git you could ask for :smile:

NielsRogge commented:
@LysandreJik I have improved the parsing of numeric values of both the question and table in `tokenization_tapas.py`. Besides this, I have fixed some of the comments I had on [my review of your PR](https://github.com/huggingface/transformers/pull/8482), and commented "done" on the ones that are fixed.

The most important things to fix now remain:

- [ ] Add correct implementation of `prev_label_ids` in case of a batch of table-question pairs (in case of a batch, all questions should refer to the same table). The implementation should reflect the [original implementation](https://github.com/google-research/tapas/blob/d8638f0909b3de32a85fe7491769d47d645d8e22/tapas/utils/tf_example_utils.py#L1155) as follows: for a given table-question pair in a batch, 

```
prev_label_ids = self.get_answer_ids(
                column_ids, row_ids, table_data, answer_text, answer_coordinates
                )
```

Here, it's important that the `get_answer_ids` function is called with the `column_ids` and `row_ids` of the **current** table-question pair in the batch, but the `answer_text` and `answer_coordinates` of the **previous** table-question pair in the batch. 

- [ ] Fix the error that I'm currently having in the colabs above (see first message of this PR), when `answer_coordinates` and `answer_text` are provided to the tokenizer
- [ ] Add support for the `drop_rows_to_fit` and `cell_trim_length` attributes of `TapasTokenizer`, which should reflect the original API

LysandreJik reviewed (commented):
Ok, I just took a look and things are starting to look very clean. I didn't see anything wrong with the question answering loss computation, but the code is so dense that it wouldn't be simple to see right away. 
Until we get an answer from the Google team, @patrickvonplaten @sgugger and @mfuntowicz it would be great if you could review and if we could get this in a near-to-merge state.

@NielsRogge if you want I can take care of the remaining steps for the tokenizer:

> - Add support for the drop_rows_to_fit and cell_trim_length attributes of TapasTokenizer, which should reflect the original API (see also my suggestion on how this could be done for cell_trim_length). Also, setting truncation=True in TapasTokenizer doesn't do anything currently.
> - I've added support for the special [EMPTY] token for empty cells in a table (based on the format_text method, see here). Does this have implications for the add_special_tokens method? I assume not? What about get_special_tokens_mask? To be verified.

Let me know if there are some remaining things where we can help out.

NielsRogge commented:
> @NielsRogge if you want I can take care of the remaining steps for the tokenizer:
> 
> > * Add support for the drop_rows_to_fit and cell_trim_length attributes of TapasTokenizer, which should reflect the original API (see also my suggestion on how this could be done for cell_trim_length). Also, setting truncation=True in TapasTokenizer doesn't do anything currently.
> > * I've added support for the special [EMPTY] token for empty cells in a table (based on the format_text method, see here). Does this have implications for the add_special_tokens method? I assume not? What about get_special_tokens_mask? To be verified.

@LysandreJik ok that would be great!

patrickvonplaten commented on docs/source/model_doc/tapas.rst:
```diff
@@ -0,0 +1,381 @@
+
+The original code can be found `here <https://github.com/google-research/tapas>`__.
+
+Tips:
```
Awesome that you guys added tips!

patrickvonplaten commented on docs/source/model_doc/tapas.rst:
```diff
@@ -0,0 +1,381 @@
+
+.. _model hub: https://huggingface.co/models?search=tapas
+
+There are 3 different ways in which one can fine-tune an already fine-tuned :class:`~transformers.TapasForQuestionAnswering` 
```
do we need the `an already fine-tuned` phrase here?

patrickvonplaten replied:
yeah it seems we need it! Never mind!

patrickvonplaten commented on docs/source/model_doc/tapas.rst:
```diff
@@ -0,0 +1,381 @@
+
+The tables themselves should be present in a folder, each table being a separate csv file. Note that the authors of the TAPAS algorithm used conversion 
+scripts with some automated logic to convert the other datasets (WTQ and WikiSQL) into the SQA format. The author explains this `here`_. This is 
+actually interesting, because these conversion scripts are not perfect (the ``answer_coordinates`` and ``float_answer`` fields are populated based on the ``answer_text``), 
```
```suggestion
Interestingly, these conversion scripts are not perfect ( ....) meaning that ...
actually interesting, because these conversion scripts are not perfect (the ``answer_coordinates`` and ``float_answer`` fields are populated based on the ``answer_text``), 
```

patrickvonplaten commented on docs/source/model_doc/tapas.rst:
```diff
@@ -0,0 +1,381 @@
+==========================================================================================
+
+Third, given that you've prepared your data in this TSV/CSV format (and corresponding CSV files containing the tabular data), you can then 
+use :class:`~transformers.TapasTokenizer` to convert table-question pairs into input_ids, attention_mask, token_type_ids and so on. Again, 
```
```suggestion
use :class:`~transformers.TapasTokenizer` to convert table-question pairs into :obj:`input_ids`, :obj:`attention_mask`, :obj:`token_type_ids` and so on. Again, 
```

patrickvonplaten commented on docs/source/model_doc/tapas.rst:
```diff
@@ -0,0 +1,381 @@
+
+Third, given that you've prepared your data in this TSV/CSV format (and corresponding CSV files containing the tabular data), you can then 
+use :class:`~transformers.TapasTokenizer` to convert table-question pairs into input_ids, attention_mask, token_type_ids and so on. Again, 
+based on which of the three cases you picked above, :class:`~transformers.TapasForQuestionAnswering` needs different things to be fine-tuned:
```
```suggestion
based on which of the three cases you picked above, :class:`~transformers.TapasForQuestionAnswering` requires different inputs to be fine-tuned:
```

patrickvonplaten commented on docs/source/model_doc/tapas.rst:
```diff
@@ -0,0 +1,381 @@
+        'numeric_values': tensor([[ ... ]]), 'numeric_values_scale: tensor([[ ... ]]), label_ids: tensor([[ ... ]])}
+
+Note that :class:`~transformers.TapasTokenizer` expects the data of the table to be text-only. You can use ``.astype(str)`` on a dataframe to turn it into
+text-only data. Of course, this only shows how to encode a single training example. In reality, you should create a PyTorch dataset and a corresponding dataloader:
```
```suggestion
text-only data. Of course, this only shows how to encode a single training example. It is advised to create a PyTorch dataset and a corresponding dataloader:
```

patrickvonplaten commented on docs/source/model_doc/tapas.rst:
```diff
@@ -0,0 +1,381 @@
+Usage: inference
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Here we explain how you can perform inference (i.e. making predictions on new data) with :class:`~transformers.TapasForQuestionAnswering`.
```
```suggestion
Here we explain how you can use :class:`~transformers.TapasForQuestionAnswering` for inference (i.e. making predictions on new data)
```

patrickvonplaten commented on docs/source/model_doc/tapas.rst:
```diff
@@ -0,0 +1,381 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Here we explain how you can perform inference (i.e. making predictions on new data) with :class:`~transformers.TapasForQuestionAnswering`.
+Basically, for inference, you only need to provide ``input_ids``, ``attention_mask`` and ``token_type_ids`` (which you can obtain using 
```
For inference, only .... have to be provided to the model to obtain output logits.

patrickvonplaten commented on docs/source/model_doc/tapas.rst:
```diff
@@ -0,0 +1,381 @@
+Here we explain how you can perform inference (i.e. making predictions on new data) with :class:`~transformers.TapasForQuestionAnswering`.
+Basically, for inference, you only need to provide ``input_ids``, ``attention_mask`` and ``token_type_ids`` (which you can obtain using 
+:class:`~transformers.TapasTokenizer`) to the model to obtain the logits. Next, you can use the handy ``convert_logits_to_predictions`` method
+of :class:`~transformers.TapasTokenizer` to convert these into actual predicted coordinates and aggregation indices. 
```
```suggestion
of :class:`~transformers.TapasTokenizer` to convert these into predicted coordinates and aggregation indices. 
```

patrickvonplaten commented on docs/source/model_doc/tapas.rst:
```diff
@@ -0,0 +1,381 @@
+TAPAS
```
Amazing doc!!!

patrickvonplaten commented on docs/source/model_doc/tapas.rst:
```diff
@@ -0,0 +1,381 @@
+        >>> answer_coordinates = [[(0, 0)], [(1, 0)], [(0, 2), (1, 2), (2, 2)]]
+        >>> answer_text = [["Brad Pitt"], ["69"], ["209"]]
+        >>> table = pd.Dataframe(data)
+        >>> inputs = tokenizer(table=table, queries=queries, answer_coordinates=answer_coordinates, answer_text=answer_text, return_tensors='pt')
```
I like the API

patrickvonplaten commented on docs/source/model_doc/tapas.rst:
```diff
@@ -0,0 +1,381 @@
+        >>> queries = ["What is the name of the first actor?", "How many movies has George Clooney played in?", "What is the total number of movies?"]
+        >>> table = pd.Dataframe(data)
+        >>> inputs = tokenizer(table, queries, return_tensors='pt')
+        >>> logits, logits_agg = model(**inputs)
```
this looks like it would break given that `use_cache=True` by default now

NielsRogge replied:
You mean `return_dict=True`?

NielsRogge replied:
Correct? @patrickvonplaten

patrickvonplaten commented on docs/source/model_doc/tapas.rst:
```diff
@@ -0,0 +1,381 @@
+        >>> data = {'Actors': ["Brad Pitt", "Leonardo Di Caprio", "George Clooney"], 'Number of movies': ["87", "53", "69"]}
+        >>> queries = ["What is the name of the first actor?", "How many movies has George Clooney played in?", "What is the total number of movies?"]
+        >>> table = pd.Dataframe(data)
+        >>> inputs = tokenizer(table, queries, return_tensors='pt')
```
no need for keyword arguments here?

NielsRogge replied:
Fixed this, should indeed have keyword arguments + padding up to max length.

patrickvonplaten commented on src/transformers/configuration_tapas.py:
```diff
@@ -0,0 +1,209 @@
+            The maximum sequence length that this model might ever be used with. Typically set this to something large
+            just in case (e.g., 512 or 1024 or 2048).
+        type_vocab_sizes (:obj:`List[int]`, `optional`, defaults to [3, 256, 256, 2, 256, 256, 10]):
+            The vocabulary sizes of the :obj:`token_type_ids` passed when calling :class:`~transformers.TapasModel`.
```
don't really understand what `type_vocab_sizes` is here... also can we link to `token_type_ids` in the glossary somehowL

NielsRogge replied:
A big difference between BERT and TAPAS is that the latter uses 7 different token type embeddings rather than just 1. So for each token type, you have a different vocabulary size. 

BERT only has the ids indicating whether a token is part of the first or second sequence (also called segment ids). TAPAS has, besides segment ids, also row ids, column ids, numeric relation ids, prev_label_ids, and column ranks. See docs of `TapasTokenizer`.

NielsRogge replied:
Ok? @patrickvonplaten

patrickvonplaten commented on src/transformers/configuration_tapas.py:
```diff
@@ -0,0 +1,209 @@
+        huber_loss_delta: (:obj:`float`, `optional`, defaults to None):
+            Delta parameter used to calculate the regression loss.
+        temperature: (:obj:`float`, `optional`, defaults to 1.0):
+            Scales cell logits to control the skewness of probabilities.
```
```suggestion
            Value used to control (OR change) the skewness of cell logits probabilities.
```

patrickvonplaten commented on src/transformers/configuration_tapas.py:
```diff
@@ -0,0 +1,209 @@
+            Importance weight for the regression loss.
+        use_normalized_answer_loss (:obj:`bool`, `optional`, defaults to :obj:`False`):
+            Normalize loss by max of predicted and expected value.
+        huber_loss_delta: (:obj:`float`, `optional`, defaults to None):
```
We can delete all `defaults to None` statements. In case a value defaults to `None` we just leave it blank -> `huber_loss_delta: (:obj:`float`, `optional`):`

patrickvonplaten replied:
should be applied to all config parameters

NielsRogge replied:
Ok

patrickvonplaten commented on src/transformers/configuration_tapas.py:
```diff
@@ -0,0 +1,209 @@
+        disable_per_token_loss: (:obj:`bool`, `optional`, defaults to :obj:`False`):
+            Disable any (strong or weak) supervision on cells.
+        span_prediction: (:obj:`string`, `optional`, defaults to :obj:`"none"`):
+            Span selection mode to use. Currently only "none" is supported.
```
Is there maybe a better description as `"none"` for the span prediction mode?

NielsRogge replied:
I think I'll remove this parameter, since it's not used.

patrickvonplaten commented on src/transformers/configuration_tapas.py:
```diff
@@ -0,0 +1,209 @@
+        allow_empty_column_selection: (:obj:`bool`, `optional`, defaults to :obj:`False`):
+            Allow not to select any column.
+        init_cell_selection_weights_to_zero: (:obj:`bool`, `optional`, defaults to :obj:`False`):
+            Whether to initialize cell selection weights to 0 so that the initial probabilities are 50%.
```
(nit) Some boolean params start with `Whether ...`, some don't . Maybe start all boolean params with `Whether`

patrickvonplaten commented on src/transformers/configuration_utils.py:
```diff
@@ -163,7 +163,7 @@ class PretrainedConfig(object):
     def __init__(self, **kwargs):
         # Attributes with defaults
-        self.return_dict = kwargs.pop("return_dict", False)
+        self.return_dict = kwargs.pop("return_dict", True)
```
this should not be shown in this PR here I think -> isn't that already changed? Maybe rebase to master?

NielsRogge replied:
Yes I know. I did this to reflect the current API of Transformers (v4). I'm going to open a clean PR with all the added files. This branch is out-of-date and a rebase would be kind of painful.

patrickvonplaten commented on src/transformers/modeling_tapas.py:
```diff
@@ -0,0 +1,2265 @@
+            attention_scores = attention_scores + attention_mask
+
+        # Normalize the attention scores to probabilities.
+        attention_probs = nn.Softmax(dim=-1)(attention_scores)
```
(nit) using functional softmax is better IMO (no object has to be instantiated...)

patrickvonplaten replied:
well it's the same in Bert I just saw, so nvm

patrickvonplaten commented on src/transformers/modeling_tapas.py:
```diff
@@ -0,0 +1,2265 @@
+        token_type_embedding_name = "token_type_embeddings"
+
+        for i in range(self.number_of_token_type_embeddings):
+            name = "%s_%d" % (token_type_embedding_name, i)
```
f-string would be prefered here I think `f"{token_type_embedding_name}_{i}"`

patrickvonplaten commented on src/transformers/modeling_tapas.py:
```diff
@@ -0,0 +1,2265 @@
+        position_ids=None,
+        head_mask=None,
+        inputs_embeds=None,
+        encoder_hidden_states=None,
```
`encoder_hidden_staets` and `encoder_attention_mask` are added. Does the model work out of the box in an encoder-decoder setting? Actually not sure whether cross attention would work here...In case cross-attention would work, it'd be great to add `encoder_hidden_states` and `encoder_attention_mask` to the docsting. I kind of doubt though that it would work so in this case I think we can delete all the `self.conif.is_decoder` related code and set:

```python
encoder_hidden_states=None,
encoder_attention_mask=None
```

further below

NielsRogge replied:
Done

patrickvonplaten commented on src/transformers/modeling_tapas.py:
```diff
@@ -0,0 +1,2265 @@
+            embedding_output,
+            attention_mask=extended_attention_mask,
+            head_mask=head_mask,
+            encoder_hidden_states=encoder_hidden_states,
```
remove this and following line IMO

patrickvonplaten commented on src/transformers/modeling_tapas.py:
```diff
@@ -0,0 +1,2265 @@
+        position_ids=None,
+        head_mask=None,
+        inputs_embeds=None,
+        encoder_hidden_states=None,
```
same if TAPAS works for encoder-decodre - add docstring for `encoder_hidden_states`, if not delete the input

patrickvonplaten commented on src/transformers/modeling_tapas.py:
```diff
@@ -0,0 +1,2265 @@
+        )
+
+
+class TapasLMHead(nn.Module):
```
add copy from BERT here as well?

patrickvonplaten commented on src/transformers/modeling_tapas.py:
```diff
@@ -0,0 +1,2265 @@
+            Aggregation function index for every example in the batch for computing the aggregation loss. Indices
+            should be in :obj:`[0, ..., config.num_aggregation_labels - 1]`. Only required in case of strong
+            supervision for aggregation (WikiSQL-SUPERVISED).
+        float_answer (:obj:`torch.FloatTensor` of shape :obj:`(batch_size, )`, `optional`):
```
I think it's better to call it `answer` instead of `float_answer`. When looking at the type one will know that it's a float

NielsRogge replied:
I actually renamed it from `answer` to `float_answer` to reflect the same column name of the SQA format. Can I leave it like that?

patrickvonplaten replied:
yeah it's fine for me!

patrickvonplaten commented on src/transformers/modeling_tapas.py:
```diff
@@ -0,0 +1,2265 @@
+        table_mask (:obj:`torch.LongTensor` of shape :obj:`(batch_size, seq_length)`, `optional`):
+            Mask for the table. Indicates which tokens belong to the table (1). Question tokens, table headers and
+            padding are 0.
+        label_ids (:obj:`torch.LongTensor` of shape :obj:`(batch_size, seq_length)`, `optional`):
```
we usually have the input `labels` -> could we maybe use this one here as well instead of `label_ids`?

NielsRogge replied:
The reason these are called `label_ids` is that one of the 7 token type ids which the authors of TAPAS define is called `prev_label_ids`. These indicate which tokens were part of an answer to the previous table-question pair, which is useful in a conversational set-up. Using `_ids` at the end reflects that these are similar to `input_ids` (also having shape `(batch_size, sequence_length)`, but are zero or one depending on whether a token is (part of) an answer. So IMO it's more clear to use `label_ids` rather than `labels`, otherwise people might be confused when they see `prev_label_ids` in the docs. 

Labels which are really similar to other `labels` in the Transformers library are the `aggregation_labels` to train `TapasForQuestionAnswering` and `labels` to train `TapasForSequenceClassification`.

patrickvonplaten replied:
From reading the description, I still think that `label_ids` is used to compute a loss. Most other `labels` in the library also have the shape `(batch_size, seq_length)` and there is no `_ids` in the end. But I'm not enough into the code to fully understand what's going on...maybe @LysandreJik can decide here :-)

NielsRogge replied:
We renamed this to labels ;) cc @patrickvonplaten

patrickvonplaten commented on src/transformers/modeling_tapas.py:
```diff
@@ -0,0 +1,2265 @@
+        # Construct indices for the table.
+        if token_type_ids is None:
+            token_type_ids = torch.zeros(
+                (*input_shape, len(self.config.type_vocab_sizes)), dtype=torch.long, device=device
```
(nit) think `input_shape + (len(...),)` is more readable

patrickvonplaten commented on src/transformers/modeling_tapas.py:
```diff
@@ -0,0 +1,2265 @@
+        Creates an index
+
+        Args:
+            indices (:obj:`torch.LongTensor`, same shape as `values`):
```
what are `values`? maybe re-define the shape here

patrickvonplaten commented on src/transformers/modeling_tapas.py:
```diff
@@ -0,0 +1,2265 @@
+    return aggregate_mask
+
+
+def _calculate_aggregation_loss_known(logits_aggregation, aggregate_mask, aggregation_function_id, config):
```
think it's better to just have `use_answer_as_supervision` and `num_aggregation_labels` as inputs instead of whole config

NielsRogge replied:
Fixed

patrickvonplaten commented on src/transformers/modeling_tapas.py:
```diff
@@ -0,0 +1,2265 @@
+    return -torch.log(aggregation_ops_total_mass) * aggregate_mask
+
+
+def _calculate_aggregation_loss(logits_aggregation, aggregate_mask, aggregation_function_id, config):
```
same no need to have whole config as input IMO

NielsRogge replied:
Fixed this, not sure if it's more readable though.

patrickvonplaten commented on src/transformers/modeling_tapas.py:
```diff
@@ -0,0 +1,2265 @@
+        average_result = torch.sum(numeric_values_masked * scaled_probability_per_cell / ex, dim=1)
+    elif avg_approximation == AverageApproximationFunction.SECOND_ORDER:
+        # The sum of all probabilities exept that correspond to other cells
+        ex = torch.sum(scaled_probability_per_cell, dim=1, keepdim=True) - scaled_probability_per_cell + 1
```
what is `ex`?

eisenjulian replied:
Hi, sorry for intruding, this is awesome! `ex` stands for expectation, more explicitly the expectation of the sum of N-1 Bernoulli random variables plus the constant 1, which is computed as adding all N expected values and subtracting the extra one. It corresponds to X_c in Appendix D in https://arxiv.org/avs/2004.02349 which is trying to approximate the average of a random set.

NielsRogge replied:
Thank you! I'll add this reference as a comment in the code.

patrickvonplaten commented on src/transformers/tokenization_tapas.py:
```diff
@@ -0,0 +1,2666 @@
+            return [1] + ([0] * len(token_ids_0)) + [1] + ([0] * len(token_ids_1))
+        return [1] + ([0] * len(token_ids_0)) + [1]
+
+    def __call__(
```
I like the API

patrickvonplaten commented on tests/test_modeling_tapas.py:
```diff
@@ -0,0 +1,612 @@
+        result = utils.gather(values, index)
+
+        # We use np.testing.assert_array_equal rather than Tensorflow's assertAllEqual
+        np.testing.assert_array_equal(result.numpy(), [[[1, 2], [3, 4]], [[7, 8], [5, 6]]])
```
I'd really like to have at least one full slow integration test here with a pretrained model

patrickvonplaten reviewed (commented):
PR looks very clean already! Great job! I left a couple of comments regarding naming, etc..
I'd really like a slow integration test and I think we should delete thee encoder-decoder related logic from `modeling_tapas.py` for now

sgugger commented on src/transformers/tokenization_tapas.py:
```diff
@@ -0,0 +1,2666 @@
+""" BasicTokenizer and WordPieceTokenizer (taken from tokenization_bert.py)"""
+
+
+class BasicTokenizer(object):
```
Can we add a comment `# Copied from transformers.models.bert.tokenization_bert.BasicTokenizer` if this is a perfect copy? That way our internal tooling will check the copy will not diverge from the original.

sgugger commented on src/transformers/tokenization_tapas.py:
```diff
@@ -0,0 +1,2666 @@
+        return "".join(output)
+
+
+class WordpieceTokenizer(object):
```
Same, can we add `# Copied from transformers.models.bert.tokenization_bert.WordpieceTokenizer`?

sgugger commented on src/transformers/tokenization_tapas.py:
```diff
@@ -0,0 +1,2666 @@
+        return output_tokens
+
+
+"""
+    Below: utilities for TAPAS tokenizer (independent from PyTorch/Tensorflow).
+
+    This includes functions to parse numeric values (dates and numbers) from both the table and questions in order
+    to create the column_ranks, inv_column_ranks, numeric_values, numeric values_scale and numeric_relations in
+    prepare_for_model of TapasTokenizer. 
+
+    These are meant to be used in an academic setup, for production use cases Gold mine or Aqua should be used.
+
+"""
```
Nit: use comments and not a docstring hee.

sgugger commented on src/transformers/tokenization_tapas.py:
```diff
@@ -0,0 +1,2666 @@
+    #### End of everything related to converting logits to predictions ####
+
+
+""" BasicTokenizer and WordPieceTokenizer (taken from tokenization_bert.py)"""
```
Can be removed if the copied comments are added below.

sgugger commented on src/transformers/tokenization_tapas.py:
```diff
@@ -0,0 +1,2666 @@
+"""
+
+
+# taken from constants.py of the original implementation
```
Link to the original implementation.

sgugger commented on src/transformers/tokenization_tapas.py:
```diff
@@ -0,0 +1,2666 @@
+    numeric_spans: Optional[List[NumericValueSpan]] = None
+
+    
+"""   
+    Below: all functions from number_utils.py as well as 2 functions (namely get_all_spans and normalize_for_match)
+    from text_utils.py of the original implementation.
+
+    URL's: 
+    - https://github.com/google-research/tapas/blob/master/tapas/utils/number_utils.py
+    - https://github.com/google-research/tapas/blob/master/tapas/utils/text_utils.py
+    
+"""
```
Nit: please use comments here.

sgugger commented on src/transformers/tokenization_tapas.py:
```diff
@@ -0,0 +1,2666 @@
+    return numeric_value_spans
+
+
+"""
+    Below: all functions from number_annotation_utils.py and 2 functions (namely filter_invalid_unicode
+    and filter_invalid_unicode_from_table) from text_utils.py of the original implementation.
+    
+    URL's: 
+    - https://github.com/google-research/tapas/blob/master/tapas/utils/number_annotation_utils.py
+    - https://github.com/google-research/tapas/blob/master/tapas/utils/text_utils.py 
+
+"""
```
Same as above

sgugger commented on src/transformers/tokenization_tapas.py:
```diff
@@ -0,0 +1,2666 @@
+    """Parses text in table column-wise and adds the consolidated values.
+    Consolidation refers to finding values with a common types (date or number).
+    Args:
+    table: Table to annotate.
+    min_consolidation_fraction: Fraction of cells in a column that need to have
+        consolidated value.
+    debug_info: Additional information used for logging.
```
Bad indent here in the docstring.

sgugger commented on tests/test_modeling_tapas.py:
```diff
@@ -0,0 +1,612 @@
+
+    # @require_torch_and_cuda
+    # def test_large_inputs_in_fp16_dont_cause_overflow(self):
+    #     pass
```
Either implement the tests above or remove the comments.

sgugger commented on tests/test_tokenization_tapas.py:
```diff
@@ -0,0 +1,1195 @@
+
+        return output_txt, output_ids
+
+    # def get_clean_sequence(self, tokenizer, with_prefix_space=False, max_length=20, min_length=5) -> Tuple[str, list]:
+    #     data = {
+    #         'Actors': ["Brad Pitt", "Leonardo Di Caprio", "George Clooney"],
+    #         'Age': ["56", "45", "59"],
+    #         'Number of movies': ["87", "53", "69"],
+    #         'Date of birth': ["18 december 1963", "11 november 1974", "6 may 1961"]
+    #     }
+    #     table = pd.DataFrame.from_dict(data)
+    #     output_ids = tokenizer.encode(table, add_special_tokens=False, max_length=max_length)
+    #     output_txt = tokenizer.decode(output_ids)
+    #
+    #     return output_txt, output_ids
```
If it's not necessary anymore, this should be removed.

sgugger commented on tests/test_tokenization_tapas.py:
```diff
@@ -0,0 +1,1195 @@
+        # Example taken from the issue https://github.com/huggingface/tokenizers/issues/340
+        self.assertListEqual([tokenizer.tokenize(t) for t in ["Test", "\xad", "test"]], [["[UNK]"], [], ["[UNK]"]])
+
+        # self.assertListEqual(
+        #     [rust_tokenizer.tokenize(t) for t in ["Test", "\xad", "test"]], [["[UNK]"], [], ["[UNK]"]]
+        # )
```
If this is not necessary anymore, this should be removed (same for lime 305).

sgugger reviewed (approved):
Wow that's a lot of work! Thanks a lot for putting this together with @LysandreJik help. This seems on track to be merged, there is still a bit of cleanup work to do before though.

NielsRogge commented on src/transformers/tokenization_tapas.py:
```diff
@@ -0,0 +1,2666 @@
+            for token, token_index in sorted(self.vocab.items(), key=lambda kv: kv[1]):
+                if index != token_index:
+                    logger.warning(
+                        "Saving vocabulary to {}: vocabulary indices are not consecutive."
+                        " Please check that the vocabulary is not corrupted!".format(vocab_file)
```
Ok, I assume I should also add an "f" before the string

NielsRogge commented on src/transformers/modeling_tapas.py:
```diff
@@ -0,0 +1,2274 @@
+        if table_mask is None:
+            table_mask = torch.where(row_ids > 0, torch.ones_like(row_ids), torch.zeros_like(row_ids))
+        # torch.FloatTensor[batch_size, seq_length] there's probably a more elegant way to do the 4 lines below
+        input_mask_float = attention_mask.type(torch.FloatTensor).to(device)
+        table_mask_float = table_mask.type(torch.FloatTensor).to(device)
```
They are special float versions of the long tensors which are used in the calculation of the regression loss.

Ok I'll use `.float()` instead. The reason I used `.to(device)` is because these copies should be on the same device as the `attention_mask` and `table_mask`. As these are copies, this is not device placement of tensors passed by the user, right?

NielsRogge commented:
Hey @LysandreJik, could you maybe fix the remaining comments? In short:

* remove the encoder-decoder logic of `TapasModel` (remove from the API, but leave them in the code that's copied from BERT (that the user won't see and won't use). I'll let you do this since I don't want to mess up anything.
* remove some tests and add a slow test as requested above

... then I'll mark these as resolved. Besides these, there's also the truncation of `TapasTokenizer` which should still be implemented.

And then I'll assume we're done (finally)

LysandreJik commented:
Sure, will do so. Probably tomorrow morning/afternoon!

NielsRogge commented on src/transformers/modeling_tapas.py:
```diff
@@ -0,0 +1,2265 @@
+
+        # If a 2D ou 3D attention mask is provided for the cross-attention
+        # we need to make broadcastabe to [batch_size, num_heads, seq_length, seq_length]
+        if self.config.is_decoder and encoder_hidden_states is not None:
```
Done

NielsRogge commented:
Closing this one as the most up-to-date is now #9117 .
